modules/rostests/apitests/gdi32/StretchBlt.c

```diff
@@ -17,92 +17,808 @@
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             1, 0, -2, 2, 0, 1, 2, -2, expected, __LINE__); // Case 0110 (6) - Both Flip.
+
+    expected[0] = 0x000000FF;
+    expected[1] = 0x0000FF00;
     expected[2] = 0x00FF0000;
+    expected[3] = 0xFF000000;
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             2, 0, -2, 2, 2, 0, -2, 2, expected, __LINE__); // Case 0101 (5) - No Flip. Special Case.
+
+    expected[0] = 0x00FF0000;
+    expected[1] = 0xFF000000;
+    expected[2] = 0x000000FF;
+    expected[3] = 0x0000FF00;
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             1, 0, -2, 2, 1, 1, -2, -2, expected, __LINE__); // Case 0100 (4) - Vertical Flip.
+
+    expected[0] = 0xFF000000;
+    expected[1] = 0x00FF0000;
+    expected[2] = 0x0000FF00;
     expected[3] = 0x000000FF;
 
-    ok(expected[1] == dstBuffer[1], "StretchBlt expected { %08X } got { %08X }\n",
-        expected[1], dstBuffer[1]);
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             1, 1, -2, -2, 0, 0, 2, 2, expected, __LINE__); // Case 0011 (3) - Both Flip. 
+
+    expected[0] = 0x0000FF00;
+    expected[1] = 0x000000FF;
+    expected[2] = 0xFF000000;
+    expected[3] = 0x00FF0000;
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             1, 1, -2, -2, 0, 1, 2, -2, expected, __LINE__); // Case 0010 (2) - Horizontal Flip. 
+
+    expected[0] = 0x00FF0000;
+    expected[1] = 0xFF000000;
+    expected[2] = 0x000000FF;
+    expected[3] = 0x0000FF00;
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             1, 1, -2, -2, 1, 0, -2, 2, expected, __LINE__); // Case 0001 (1) - Vertical Flip. 
+
+    expected[0] = 0x000000FF;
+    expected[1] = 0x0000FF00;
+    expected[2] = 0x00FF0000;
+    expected[3] = 0xFF000000;
+
+    check_StretchBlt_stretch(hdcDst, hdcSrc, &biDst, dstBuffer, srcBuffer,
+                             2, 2, -2, -2, 2, 2, -2, -2, expected, __LINE__); // Case 0000 (0) - No Flip. Special Case.
 
     DeleteDC(hdcSrc);
+
+    SelectObject(hdcDst, oldDst);
+    DeleteObject(bmpDst);
     DeleteDC(hdcDst);
+
+    DeleteDC(hdcScreen);
+}
+
+static void test_StretchBlt_TopDownOptions(BOOL SrcTopDown, BOOL DstTopDown)
+{
+    HBITMAP bmpDst, bmpSrc;
+    HBITMAP oldDst, oldSrc;
+    HDC hdcScreen, hdcDst, hdcSrc;
+    UINT32 *dstBuffer, *srcBuffer;
+    BITMAPINFO biDst, biSrc;
+    UINT32 expected[256];
+
+    memset(&biDst, 0, sizeof(BITMAPINFO));
+
+    biDst.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+    biDst.bmiHeader.biWidth = 2;
+    biDst.bmiHeader.biHeight = 2;
+    biDst.bmiHeader.biPlanes = 1;
+    biDst.bmiHeader.biBitCount = 32;
+    biDst.bmiHeader.biCompression = BI_RGB;
+    memcpy(&biSrc, &biDst, sizeof(BITMAPINFO));
+
```

Comment:
I can see inconsistency of the relation of *TopDown flag and biHeight around here and below .

Reply:
You can have a Top Down surface and use it to render a Top Down or Bottom Up bitmap. I do not think that there is any inconsistency here. Please provide more details if you believe otherwise. Thanks. Doug Lyons

Reply:
[ft3.zip](https://github.com/reactos/reactos/files/5979818/ft3.zip)
@katahiromz, you may want to review the attached program regarding this. Thanks.